src/main/java/de/tum/cit/aet/artemis/programming/service/localci/LocalCIResultProcessingService.java

```diff
@@ -283,36 +280,34 @@ private BuildJob saveFinishedBuildJob(BuildJobQueueItem queueItem, BuildStatus b
-
-            boolean hasSuccessfulBuildJobs = successfulBuildJobCountByExerciseId > 0;
-            boolean exceedsUpdateThreshold = buildStatistics == null
-                    || successfulBuildJobCountByExerciseId - buildStatistics.getBuildCountWhenUpdated() >= BUILD_STATISTICS_UPDATE_THRESHOLD;
-
-            boolean shouldUpdate = hasSuccessfulBuildJobs && exceedsUpdateThreshold;
-            if (!shouldUpdate) {
+            var buildStatisticsDto = buildJobRepository.findBuildJobStatisticsByExerciseId(exerciseId);
+            if (buildStatisticsDto == null || buildStatisticsDto.buildCountWhenUpdated() == 0) {
                 return;
             }
+            var programmingExerciseBuildStatistics = programmingExerciseBuildStatisticsRepository.findByExerciseId(exerciseId).orElse(null);
 
-            OptionalDouble averageBuildDuration = buildJobRepository.fetchSuccessfulBuildJobsByExerciseIdWithLimit(exercise.getId(), BUILD_JOB_DURATION_UPDATE_LIMIT).stream()
-                    .mapToLong(buildJob -> Duration.between(buildJob.getBuildStartDate(), buildJob.getBuildCompletionDate()).toSeconds()).average();
-            if (averageBuildDuration.isPresent()) {
-                if (buildStatistics == null) {
-                    buildStatistics = new ProgrammingExerciseBuildStatistics(exercise.getId(), Math.round(averageBuildDuration.getAsDouble()), successfulBuildJobCountByExerciseId);
-                }
-                else {
-                    buildStatistics.setBuildDurationSeconds((long) averageBuildDuration.getAsDouble());
-                    buildStatistics.setBuildCountWhenUpdated(successfulBuildJobCountByExerciseId);
+            long averageDuration = Math.round(buildStatisticsDto.buildDurationSeconds());
+
+            if (programmingExerciseBuildStatistics == null) {
+                // create the database row if it does not exist
+                programmingExerciseBuildStatistics = new ProgrammingExerciseBuildStatistics(exerciseId, averageDuration, buildStatisticsDto.buildCountWhenUpdated());
+                programmingExerciseBuildStatisticsRepository.save(programmingExerciseBuildStatistics);
+            }
+            else {
+                // only update the database row if the build duration has changed using a modifying query or when the build count is above a certain threshold
+                if (averageDuration == programmingExerciseBuildStatistics.getBuildDurationSeconds()
+                        && buildStatisticsDto.buildCountWhenUpdated() - programmingExerciseBuildStatistics.getBuildCountWhenUpdated() < BUILD_STATISTICS_UPDATE_THRESHOLD) {
+                    return;
                 }
-                programmingExerciseBuildStatisticsRepository.save(buildStatistics);
+                programmingExerciseBuildStatisticsRepository.updateStatistics(averageDuration, buildStatisticsDto.buildCountWhenUpdated(), exerciseId);
             }
```

Comment:
I think you could simplify the control flow here (pseudo code):
```
else if (averageDuration != programmingExerciseBuildStatistics.getBuildDurationSeconds()
                        || buildStatisticsDto.buildCountWhenUpdated() - programmingExerciseBuildStatistics.getBuildCountWhenUpdated() >= BUILD_STATISTICS_UPDATE_THRESHOLD) {
                        programmingExerciseBuildStatisticsRepository.updateStatistics(averageDuration, buildStatisticsDto.buildCountWhenUpdated(), exerciseId);
                        }
```

This way you don't have the return, and it's a bit less nested